presto-prometheus/src/main/java/io/prestosql/plugin/prometheus/PrometheusClient.java

```diff
@@ -0,0 +1,148 @@
+
+import static java.nio.charset.StandardCharsets.UTF_8;
+import static java.util.Objects.requireNonNull;
+
+public class PrometheusClient
+{
+    // schema name is fixed to "prometheus"
+    private final Supplier<Map<String, Object>> tableSupplier;
+    protected final PrometheusConfig config;
+    private static TypeManager typeManager;
+    protected static final String METRICS_ENDPOINT = "/api/v1/label/__name__/values";
+
+    @Inject
+    public PrometheusClient(PrometheusConfig config, JsonCodec<Map<String, Object>> metricCodec, TypeManager typeManager)
+            throws URISyntaxException
+    {
+        requireNonNull(config, "config is null");
+        requireNonNull(metricCodec, "metricCodec is null");
+        requireNonNull(typeManager, "typeManager is null");
+
+        tableSupplier = Suppliers.memoizeWithExpiration(metricsSupplier(metricCodec, getPrometheusMetricsURI(config)), 30, TimeUnit.SECONDS);
+        this.config = config;
+        this.typeManager = typeManager;
+    }
+
+    public Set<String> getSchemaNames()
+    {
+        return ImmutableSet.of("prometheus");
+    }
+
+    private URI getPrometheusMetricsURI(PrometheusConfig config)
+            throws URISyntaxException
+    {
+        // endpoint to retrieve metric names from Prometheus
+        URI uri = config.getPrometheusURI();
+        return new URIBuilder()
+                .setScheme(uri.getScheme())
+                .setHost(uri.getAuthority())
+                .setPath(uri.getPath().concat(METRICS_ENDPOINT))
+                .build();
+    }
+
+    public Set<String> getTableNames(String schema)
+    {
+        requireNonNull(schema, "schema is null");
+        if (schema.equals("prometheus")) {
+            String status = (String) tableSupplier.get().get("status");
+            if (status.equals("success")) {
+                List<String> tableNames = (List<String>) tableSupplier.get().get("data");
+                if (tableNames == null) {
+                    return ImmutableSet.of();
+                }
+                return tableNames.stream().collect(Collectors.toSet());
+            }
+            else {
+                //TODO this deals with success|error, there may be warnings to handle
+                return ImmutableSet.of();
+            }
+        }
+        return ImmutableSet.of();
+    }
+
+    public PrometheusTable getTable(String schema, String tableName)
+    {
+        requireNonNull(schema, "schema is null");
+        requireNonNull(tableName, "tableName is null");
+        if (schema.equals("prometheus")) {
+            List<String> tableNames = (List<String>) tableSupplier.get().get("data");
+            if (tableNames == null) {
+                return null;
+            }
+            if (!tableNames.contains(tableName)) {
+                return null;
+            }
+            PrometheusTable table = new PrometheusTable(
+                    tableName,
+                    ImmutableList.of(
```

Comment:
I wonder, can we have a single instance of this list for all tables declared somewhere else?

Reply:
This is its only usage in non-test code. In the test code we use a typemanager derived specifically for testing.

Reply:
I'm not saying to move it to make it reusable in other code, I'm saying to move it so we can avoid re-creating the same list every time this is called, and so the list object can be shared in-memory across all PrometheusTable instances.

Reply:
Ah, makes sense. Done in 48c2dbe